spyder/plugins/workingdirectory/plugin.py

```diff
@@ -110,27 +114,76 @@ def on_ipyconsole_available(self):
             ipyconsole.set_current_client_working_directory)
         # TODO: chdir_current_client might follow a better naming
         # convention
-        ipyconsole.sig_current_directory_changed.connect(
+        self._ipyconsole_chdir = (
             lambda path, plugin=ipyconsole: self.chdir(path, plugin))
+        ipyconsole.sig_current_directory_changed.connect(
+            self._ipyconsole_chdir)
 
     @on_plugin_available(plugin=Plugins.Projects)
     def on_projects_available(self):
         projects = self.get_plugin(Plugins.Projects)
-        projects.sig_project_loaded.connect(
+        self._projects_loaded = (
             lambda path:
             self.chdir(
                 directory=path,
                 sender_plugin=projects
            )
         )
-
-        projects.sig_project_closed[object].connect(
+        self._projects_closed = (
             lambda path: self.chdir(
                 directory=projects.get_last_working_dir(),
                 sender_plugin=projects
             )
         )
 
+        projects.sig_project_loaded.connect(self._projects_loaded)
+        projects.sig_project_closed[object].connect(self._projects_closed)
+
+    @on_plugin_teardown(plugin=Plugins.Toolbar)
+    def on_toolbar_teardown(self):
+        toolbar = self.get_plugin(Plugins.Toolbar)
+        toolbar.remove_application_toolbar('working_directory_toolbar')
+
+    @on_plugin_teardown(plugin=Plugins.Preferences)
+    def on_preferences_teardown(self):
+        preferences = self.get_plugin(Plugins.Preferences)
+        preferences.deregister_plugin_preferences(self)
+
+    @on_plugin_teardown(plugin=Plugins.Editor)
+    def on_editor_teardown(self):
+        editor = self.get_plugin(Plugins.Editor)
+        editor.sig_dir_opened.disconnect(self._editor_dir_chdir)
+        self._editor_dir_chdir = None

```

Comment:
Please remove this line because it wont' be necessary after this  is a proper method.